core/lib/Hooks.rei

```diff
@@ -0,0 +1,77 @@
+type hook('a) = ..;
+
+module Slots: Slots.S with type elem('a) = hook('a);
+
+type t('a, 'b) = {
+  slots: Slots.t('a, 'b),
+  onSlotsDidChange: unit => unit,
+};
+type empty = t(unit, unit);
+
+let create: (~onSlotsDidChange: unit => unit) => t('a, 'b);
+
+module State: {
+  type t('a);
+  type hook('a) +=
+    pri
+    | State(t('a)): hook(t('a));
+};
+
+module Reducer: {
+  type t('a);
+  type hook('a) +=
+    pri
+    | Reducer(t('a)): hook(t('a));
+};
+
+module Ref: {
+  type t('a);
+  type hook('a) +=
+    pri
+    | Ref(t('a)): hook(t('a));
+};
+
+module Effect: {
+  type t('a);
+  type lifecycle =
+    | Mount
+    | Unmount
+    | Update;
+  type always;
+  type onMount;
+  type condition('a) =
```

Comment:
Trying to wrap my head around this... if this is already a GADT, does the `'a` need to be exposed? Would something like this be possible?

```reason
type condition =
    | Always: condition
    | OnMount: condition
    | If(('a, 'a) => bool, 'a): condition;
```

Or alternatively, make it a regular variant / ADT.

I guess I'm not understanding why this is a GADT 😄

Reply:
Good question. The reason is that I want the constructor to carry more type information so that you cannot do this:

```
Hooks.effect(someProp ? Always : OnMount, () => None)
```

`Always` and `OnMount` have different types so the semantics here would be very confusing!

Reply:
also, for the if condition we have to know what type the compared value is!